JDK-8296226: Add constructors (String,Throwable) and (Throwable) to InvalidParameterException

MBaesken opened the pull request:
This change adds constructors (String,Throwable) and (Throwable) to InvalidParameterException and uses them at a few places in the jdk coding.
<!-- csr: 'update' -->

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8296226](https://bugs.openjdk.org/browse/JDK-8296226): Add constructors (String,Throwable) and (Throwable) to InvalidParameterException
 * [JDK-8296237](https://bugs.openjdk.org/browse/JDK-8296237): Add constructors (String,Throwable) and (Throwable) to InvalidParameterException (**CSR**)


### Reviewers
 * [Sean Mullan](https://openjdk.org/census#mullan) (@seanjmullan - **Reviewer**)
 * [Martin Doerr](https://openjdk.org/census#mdoerr) (@TheRealMDoerr - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10966/head:pull/10966` \
`$ git checkout pull/10966`

Update a local copy of the PR: \
`$ git checkout pull/10966` \
`$ git pull https://git.openjdk.org/jdk pull/10966/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10966`

View PR using the GUI difftool: \
`$ git pr show -t 10966`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10966.diff">https://git.openjdk.org/jdk/pull/10966.diff</a>

</details>

ExE-Boss commented on src/java.base/share/classes/java/security/InvalidParameterException.java:
```diff
@@ -58,4 +58,25 @@ public InvalidParameterException() {
+     *
+     * @param msg the detail message.
+     * @param cause the cause.
+     */
```
```suggestion
     * @since 20
     */
```

ExE-Boss commented on src/java.base/share/classes/java/security/InvalidParameterException.java:
```diff
@@ -58,4 +58,25 @@ public InvalidParameterException() {
+     * specified cause.
+     *
+     * @param cause the cause.
+     */
```
```suggestion
     * @since 20
     */
```

ExE-Boss reviewed (commented):
These should probably have `@since` tags:

MBaesken commented:
Hi ExE-Boss, thanks for the suggestion , I will add @since tags .

seanjmullan commented:
The new constructors are underspecified. They should be consistent with the specification of the superclass constructors, which describe the behavior of `null` parameters and give more details on the error message. I would basically copy most of the description and params from the `Throwable` constructors. See, for example the `java.lang.IllegalArgumentException` constructors that take `Throwable` params.

MBaesken commented:
Hi Sean, I adjusted the constructor javadoc related comments, also added the since tags.

seanjmullan commented on src/java.base/share/classes/java/security/InvalidParameterException.java:
```diff
@@ -58,4 +58,44 @@ public InvalidParameterException() {
+
+    /**
+     * Constructs an {@code InvalidParameterException} with the
+     * specified detail message and cause. A detail message is a {@code String} that describes
```
Keep these lines within 80 character width to be consistent with the rest of the file.

seanjmullan commented on src/java.base/share/classes/java/security/InvalidParameterException.java:
```diff
@@ -58,4 +58,44 @@ public InvalidParameterException() {
+    }
+
+    /**
+     * Constructs an {@code InvalidParameterException} with the specified cause and a detail
```
Keep these lines within 80 character width to be consistent with the rest of the file.

seanjmullan commented on src/java.base/share/classes/java/security/InvalidParameterException.java:
```diff
@@ -58,4 +58,44 @@ public InvalidParameterException() {
+     *         is permitted, and indicates that the cause is nonexistent or
+     *         unknown.)
+     *
+     * @since 20
```
Nit: align "20" with text (add a space) to be consistent with other ctor.

seanjmullan commented on src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11KeyGenerator.java:
```diff
@@ -335,8 +335,7 @@ protected void engineInit(int keySize, SecureRandom random) {
         } catch (InvalidAlgorithmParameterException iape) {
-            throw (InvalidParameterException)
-                    (new InvalidParameterException().initCause(iape));
+            throw new InvalidParameterException("Invalid algorithm parameter", iape);
```
I would change this to `throw new InvalidParameterException(iape)`. The detail message of the cause is more useful in this case.

seanjmullan commented on src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11KeyPairGenerator.java:
```diff
@@ -151,7 +151,7 @@ public void initialize(int keySize, SecureRandom random) {
             checkKeySize(keySize, null);
         } catch (InvalidAlgorithmParameterException e) {
-            throw (InvalidParameterException) new InvalidParameterException(e.getMessage()).initCause(e);
+            throw new InvalidParameterException(e.getMessage(), e);
```
Here too I would change this to `throw new InvalidParameterException(e)` since the detail message of the cause will be used, the first parameter doesn't add any value.

seanjmullan commented on test/jdk/sun/security/tools/keytool/fakegen/java.base/sun/security/rsa/RSAKeyPairGenerator.java:
```diff
@@ -57,7 +57,7 @@ public void initialize(int keySize, SecureRandom random) {
                     RSAKeyGenParameterSpec.F4), random);
         } catch (InvalidAlgorithmParameterException iape) {
-            throw new InvalidParameterException(iape.getMessage());
+            throw new InvalidParameterException(iape.getMessage(), iape);
```
Use the one-arg ctor to make this consistent with the code above: [src/java.base/share/classes/sun/security/rsa/RSAKeyPairGenerator.java](https://github.com/openjdk/jdk/pull/10966/files#diff-4eaff4fafb2999bb74081347365803821458dc8ea618808474a948a4dc7d9b75)

seanjmullan reviewed (changes requested):
Please also update test/jdk/java/security/Exceptions/ChainingConstructors.java and add a new test for these constructors.

MBaesken commented:
Hi Sean, I adjusted the javadoc and the exceptions mentioned.
Also added InvalidParameterException to the ChainingConstructors test.
(btw the whole ChainingConstructors  test coding looks a bit bloated but that is maybe for another PR)

seanjmullan reviewed (approved):
Everything LGTM now. Thanks.

MBaesken commented:
Hi Sean, thanks for the review !

TheRealMDoerr commented on test/jdk/java/security/Exceptions/ChainingConstructors.java:
```diff
@@ -185,6 +185,16 @@ public static void main(String[] args) throws Exception {
         }
 
+        InvalidParameterException ipe =
+                new InvalidParameterException(cause);
```
Indentation looks unusual, but it's consistent with some other parts. No need to change!

TheRealMDoerr reviewed (approved):
LGTM, too.

MBaesken commented:
/integrate
